### PR TITLE
add plot bootstrap coefficients

### PR DESCRIPTION
**Pull request by JoseAlanis**
Add example code for bootstraping coeffients of linear model estimator

Example output for the electrode B8 (here showing the effect of phase-coherence):
PC-beta-B8

### CODE REVIEW

File: plot_bootstrap_coefficients.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+# choose which estimator to use (either "ridge" or "ols"
+estimator = 'ridge'
+
+# run bootstrap for regression coefficients
+boot_betas = []
+for i in range(boot):
+    resamples = np.random.choice(range(n_epochs), n_epochs)
+    # set up model: ridge or ols
+    if estimator == 'ridge':
```
Comment (dengemann) [inline plot_bootstrap_coefficients.py]: The ridge is not helpful here. Can we just remove it for now?

File: plot_bootstrap_coefficients.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+# number of random samples
+boot = 2000
+
+# choose which estimator to use (either "ridge" or "ols"
+estimator = 'ridge'
+
+# run bootstrap for regression coefficients
+boot_betas = []
+for i in range(boot):
+    resamples = np.random.choice(range(n_epochs), n_epochs)
```
Comment (dengemann) [inline plot_bootstrap_coefficients.py]: Setting the seed is important.
Can you do create a state object from `np.random.RandomState` and use it's `choice` method?
Also make sure that `replace=True`. Better to set it explicitly to make the point clear.

Comment (dengemann) [reply plot_bootstrap_coefficients.py]: And you'd ideally set the seed outside the loop.

File: plot_bootstrap_coefficients.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+# -- initialize bootstrap procedure --
+# set random state for replication
+random_state = 42
+np.random.seed(random_state)
+
+# number of random samples
+boot = 2000
+
+# choose which estimator to use (either "ridge" or "ols"
+estimator = 'ridge'
```
Comment (dengemann) [inline plot_bootstrap_coefficients.py]: remove this please.

File: plot_bootstrap_coefficients.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+n_channels = len(picks_eeg)
+n_times = len(limo_epochs['2'].times)
+
+# vectorize (channel) data for linear regression
+Y = Vectorizer().fit_transform(data)
+
+# -- initialize bootstrap procedure --
+# set random state for replication
+random_state = 42
+np.random.seed(random_state)
```
Comment (dengemann) [inline plot_bootstrap_coefficients.py]: use RandomState object below.

File: plot_bootstrap_coefficients.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    linear_model.fit(X=design.iloc[resamples], y=Y[resamples, :])
+    # extract coefficients
+    boot_betas.append(get_coef(linear_model, 'coef_'))
+
+# compute low and high percentiles
+lower, upper = np.quantile(boot_betas, [.025, .975], axis=0)
+
+###############################################################################
+# set up model: ridge or ols
+if estimator == 'ridge':
```
Comment (dengemann) [inline plot_bootstrap_coefficients.py]: remove ridge.

File: plot_bootstrap_coefficients.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+# project coefficients back to a channels x time points space.
+lm_betas = dict()
+ci = dict(lower_bound=dict(), upper_bound=dict())
+# loop through predictors
+for ind, predictor in enumerate(predictors):
+    # extract coefficients for predictor in question
+    beta = betas[:, ind]
+    lower_bound = lower[:, ind]
+    upper_bound = upper[:, ind]
```
Comment (dengemann) [inline plot_bootstrap_coefficients.py]: you could do the reshaping in one line no?

**Comment by dengemann** [comment]
Other than my comments it looks clean.

**Comment by dengemann** [comment]
@JoseAlanis looking at #5 we should probably just combine these PRs into one.
They show the same, don't they?

**Comment by dengemann** [comment]
Probably you can just close this one for now and see which of my comments apply to #5, i.e., those about random state.

**Comment by JoseAlanis** [comment]
> @JoseAlanis looking at #5 we should probably just combine these PRs into one.
> They show the same, don't they?

yup, they are very similar. #5 looks at the mean (the ERP) and #6 looks at the betas. But the approach is pretty much the same.

**Comment by dengemann** [comment]
what is the difference between beta and ERP? Betas estimate the ERP.

**Comment by JoseAlanis** [comment]
> what is the difference between beta and ERP? Betas estimate the ERP.

I see your point. The PRs do show some redundant information, Originally, I thought it would be good to show people how the bootstrap procedure is done for the Betas (i.e., actually fitting the model 2000 times on random samples of the original data). But I guess once one has understood how the bootstrap is carried out for the mean it should be easy to extrapolate to the linear model. But perhaps you are right and we could just close this PR. We'll pick up this approach later in other analyses.

**Comment by dengemann** [comment]
It is really the same to me. We’re really interested in bootstrapping beta.

**Comment by JoseAlanis** [comment]
> It is really the same to me. We’re really interested in bootstrapping beta.

Alright, you're right. Closing here for now.
